sanity-io/sanity
# feat(sanity): add document comparison tool

### Description

This branch adds a diff view for comparing documents. Primarily, this is for use with versions, but it has been designed to support comparing any two documents.

Most of the commits are supporting work (this should probably be a stack). The main feature addition occurs in  4f030aa23991dabdda009283e9d648f18dd0baab.
#### Navigating to diff view

Users can navigate to the diff view by setting the following URL search parameters:

```
| Name               | Type                  | Required | Description                                                                                              |
|--------------------|-----------------------|----------|----------------------------------------------------------------------------------------------------------|
| `diffView`         | `'version'`           | Required | Enable *version comparison* mode UI.                                                                     |
| `previousDocument` | `${string},${string}` | Required | The document type and id identifying the *previous* document for comparison, delimited by `,` character. |
| `nextDocument`     | `${string},${string}` | Required | The document type and id identifying the *next* document for comparison, delimited by `,` character.     |
```

##### Example: specifying both `previousDocument` and `nextDocument`

```
/test/structure/author;8701af36-f4e2-4f32-b13e-fc522ddb381f?perspective=r64awNgUC&structure[diffView]=version&structure[nextDocument]=author%2Cversions.r64awNgUC.8701af36-f4e2-4f32-b13e-fc522ddb381f&structure[previousDocument]=author%2Cversions.rO2NP9B76.8701af36-f4e2-4f32-b13e-fc522ddb381f
```

### New dependency: [ScrollMirror](https://www.npmjs.com/package/scrollmirror)

I've added [ScrollMirror](https://www.npmjs.com/package/scrollmirror) to handle scroll position syncing between two panes that may have different heights. This is achieved by syncing the percentage scroll position of each pane. We could implement this directly in the Studio codebase if we wish.

### What to review

- General implementation.
- Introduction of `ScrollMirror` dependency.

### Testing

There are no tests just yet, as I've spent more time than I expected making the code render in a test environment. First, I [needed to move the diff view code](https://github.com/sanity-io/sanity/commit/2cc93c797f028cc172e11499de8fa9a0725495dc) that was originally organised as a plugin to be part of `sanity/structure`. Then I encountered many issues with existing imports in `sanity/structure` resolving to `undefined` when running in a test environment.

- Navigate to a diff view. Ensure the UI works as expected.
- Navigate to a diff view using the "Compare versions" document action. Ensure the diff view is opened as expected.

I will follow up with tests as soon as these problems are resolved.

### Notes for release

The new document comparison tool provides a side-by-side view of two different document by versions, highlighting the fields that have changed between the _previous_ and _next_ versions. Select "Compare versions" from the document actions menu when viewing any draft or version document in Studio to open the tool.

By default, the first pane displays the _previous_ version, and the second pane shows _next_ version (the version that introduces changes). Use the menu at the top of each pane to switch the versions displayed.
